### PR TITLE
Scipy optimize

### PR DESCRIPTION
**Pull request by michael-hartmann**
This PR uses the new optimize.minimize API also in the programs plot_gradient_descent.py and helper/compare_optimizers.py. This PR also completes the migration to the new API.

However, there is still a problem in helper/compare_optimizers.py. For the solver L-BFGS_B I get the warning
 ``OptimizeWarning: Unknown solver options: factr``.
According to the documentation, https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html, this option does exist. Unfortunately, I don't have time to investigate this issue myself.

If nobody has time to look at the problem right now, the best way would be to create a new issue, while closing the issues #215 and #300 .

### CODE REVIEW

File: advanced/mathematical_optimization/examples/helper/compare_optimizers.py [hunk 1]
```diff
@@ -15,30 +15,31 @@
...
-                                        avextol=1e-7, maxiter=5e3),
-    'Conjugate gradient':   my_partial(optimize.fmin_cg,
-                                        gtol=1e-7, maxiter=5e3),
-    'L-BFGS':               my_partial(optimize.fmin_l_bfgs_b,
-                                        approx_grad=1, factr=10.0,
-                                        pgtol=1e-8, maxfun=1e7),
-    "L-BFGS w f'":          my_partial(optimize.fmin_l_bfgs_b,
-                                        factr=10.0,
-                                        pgtol=1e-8, maxfun=1e7),
+    'Nelder-mead':          my_partial(method="Nelder-Mead", options = {
```
Comment (GaelVaroquaux) [inline advanced/mathematical_optimization/examples/helper/compare_optimizers.py]: PEP8 tells us that there should no space around the equal sign after "options" (here and below)

**Comment by GaelVaroquaux** [comment]
I think that the problem with factr is that it has been replaced by ftol. I would put "ftol=1e-13" and remove factr to avoid the warning.

Aside from this, and the PEP8 comment above, this is good to merge for me. Thanks a lot!

**Comment by GaelVaroquaux** [comment]
@michael-hartmann : will you have time to address the comments, or should someone else try to finish off this PR?

**Comment by gertingold** [comment]
@michael-hartmann is travelling and I am not sure that he will be able to react quickly. Should I finish his work?

**Comment by GaelVaroquaux** [comment]
If you think that he will come back to this in a week or two, let's leave
it for him. I want to do a release, but it can wait a week or two.

If you have spare cycles, I think that improving the contributor guide
would be useful. At the sprint, it seemed to me that the workflow wasn't
obvious for newcomers.

**Comment by gertingold** [comment]
It might be even a bit more than two weeks but there is no harm in waiting as long as it is ok for you. Completing the PR does not seem to be a big deal and it can be done relatively fast if necessary. If I find time, I will take a look at the contributor guide.

**Comment by michael-hartmann** [comment]
I hope I find some time for it withing two weeks

**Comment by michael-hartmann** [comment]
Sorry, I had some troubles while travelling and couldn't find time to make the changes earlier...

I have fixed the spaces in agreement with PEP8 and replaced factr by ftol.

**Comment by GaelVaroquaux** [comment]
LGTM. Merging. Thank you!

**Pull request merged**